feat(deployments): add concurrency support

jimid27 opened the pull request:
- Resolves: https://linear.app/prefect/issue/PLA-701/deployments-support-concurrency
- Adds concurrency limit which is an int and concurrency options which is an object
- Based on the current api docs for deployments found [here](https://docs.prefect.io/v3/api-ref/rest-api/server/deployments/create-deployment)

jimid27 committed:
initial commit for concurrency support

mitchnielsen committed:
Change ConcurrencyOptions from JSON to SingleNestedAttribute

This object in the API is pretty simple, with only one nested attribute.
It's not so flexible that we need to default to a basic JSON object.

mitchnielsen commented:
🤔 Looking into concurrency_limit more. Looks like it is deprecated in favor of concurrency_limit_id based on [this code](https://github.com/PrefectHQ/prefect/blob/7903f46743f9a55439c85e5ad78d2ebd88ef869e/src/prefect/client/schemas/filters.py#L540-L544). That change was added in https://github.com/PrefectHQ/prefect/pull/15426, which has a lot more context on the change. I need to read up on that to understand how we need to implement the ability in the Terraform provider.

The relevant docs (https://docs.prefect.io/v3/deploy/index#concurrency-limiting) still instruct you to set `concurrency_limit` directly, and the API docs don't say it's deprecated.

I'll read up more and reach out to the backend folks if anything needs clarification.

mitchnielsen committed:
Merge branch 'main' of github.com:PrefectHQ/terraform-provider-prefect into chore/support-concurrency-in-deployments

jimid27 committed:
Merge main

jimid27 committed:
Pull

mitchnielsen committed:
Validate that concurrency_limit value is >0

Per the API docs

mitchnielsen committed:
Fix concurrency limit return value

Concurrency limit is still configured via 'concurrency_limit' in the
request payload, but that field in the response payload is deprecated.
We instead need to retrieve it from 'global_concurrency_limit.limit'.

mitchnielsen committed:
Omit ConcurrencyOptions from payload if null

If ConcurrencyOptions is not configured, it should not be included in
the request payload. We get this by using pointers and checking IsNull
before attempting to add that field to the request payloads.

mitchnielsen committed:
Explain unused fields in GlobalConcurrencyLimit

mitchnielsen committed:
Mark collision_strategy as required, update docs

mitchnielsen committed:
Undo go.mod change

mitchnielsen committed:
Check if GlobalConcurrencyLimit is nil

To be safe, let's check if GlobalConcurrencyLimit is nil before
attempting to get its nested value.

Also marks ConcurrencyLimit as a pointer value for consistency.

parkedwards commented on internal/provider/resources/deployment_test.go:
```diff
@@ -22,6 +22,8 @@ type deploymentConfig struct {
 	DeploymentName         string
 	DeploymentResourceName string
 
+	ConcurrencyLimit       int
```
should we just make this an int64, since you have to cast it

parkedwards commented on internal/api/deployments.go:
```diff
@@ -75,3 +82,22 @@ type DeploymentUpdate struct {
+
+// GlobalConcurrencyLimit is a representation of the deployment global concurrency limit.
+type GlobalConcurrencyLimit struct {
+	Limit int `json:"limit"`
```
should we just make this an int64, since you're casting it

mitchnielsen replied:
good catch, updated in 599e7e0

mitchnielsen committed:
Change GlobalConcurrencyLimit.Limit to int64

Since we were casting it as such anyway

mitchnielsen committed:
Change ConcurrencyOptions to a custom type

ConcurrencyOptions was a types.Object, which made working with the
object more verbose for a resource that was a simple nested attribute.

Also, getting values from that object was causing the JSON request
payload to include escaped characters for some reason, meaning we needed
a helper function to strip them out.

This changes from types.Object to a custom Go type, allowing much easier
manipulation of the attribute - and removing the need for the custom
helper function to strip out quotes.

mitchnielsen committed:
Merge branch 'main' of github.com:PrefectHQ/terraform-provider-prefect into chore/support-concurrency-in-deployments

parkedwards reviewed (approved):
🔥
